build-system/tasks/dist.js

```diff
@@ -88,9 +88,16 @@ function transferSrcsToTempDir() {
 
 /**
  * Dist Build
+ * @param {Object|Function} args
  * @return {!Promise}
  */
-async function dist() {
+async function dist(args) {
+  // The default 'dist' task passes in a callback of type Function. The testing
+  // tasks ('e2e', 'integration', and 'visual-diff') pass in args via an Object.
+  if (typeof args == 'object') {
```

Comment:
Not familiar with this code, but do we want to overwrite `argv` here? Or does it make sense to update argv instead (using `Object.assign`)?.

It also feels weird to be modifying argv, other functions in this file use it too. It feels weird if there were somehow two calls to `dist` in the same process, since one call would affect the second (if it did not pass in args as an Object).

As a cleanup, it may make sense to stop referring to the global argv in the various functions, and pass everything that is needed to them.

Reply:
> ... do we want to overwrite `argv` here?
> ... stop referring to the global argv...

`argv` is a locally parsed copy of the global argument list which is created and used by each task, so modifying it here does not alter the global list `process.argv`.

> ... if there were somehow two calls to dist in the same process...

There is no precedent for any given `gulp` task calling `dist` twice, since it is an extremely expensive call. The only two ways in which `dist()` may be called are:
- By the `gulp dist` task (where the argument is a callback of type `Function` passed in by `gulp`)
- By a testing task like `e2e`, `integration`, or `visual-diff` that requires a pre-build (where the argument is the local `argv` copy of type `object` populated by the calling task)

Reply:
I tend to agree with @sparhami that it feels a bit weird to modify this, even with all the caveats you mentioned. It makes it a bit more difficult to follow the trail of data throughout the code as well, because readers will always assume that `argv` is constant. Any way you can make this more "procedural"?

Reply:
Fair points. I've separated the `gulp` entry point (`dist()`) into a helper function (`performDist()`), removed the `typeof` hack, and passed in extra arguments via an object.

Reply:
After some more testing, it looks like my attempts to avoid nested gulp invocations by playing with `argv` are insufficient, since it is parsed separately by each file. I'm going to accept defeat and revert the test tasks to how they were originally written, which is a call to `execOrDie('gulp dist --fortesting --config <config>')`. New commit coming up.